### PR TITLE
feat(workflow): improve workflow log readability for humans

### PR DESCRIPTION
**Pull request by reubenmiller**
## Proposed changes

Improve workflow layout and styling to make it easier and quicker for humans to follow complex workflows. The following changes were implemented:
* Display nested workflow states visually
* Display state contents in input/output format
* Add log header section to display workflow trigger info
* Remove duplicated information from each state
* Include user friendly hints for exit code and empty stdout/stderr

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by reubenmiller** [comment]
Attached is a full example log file taken from a custom workflow (trying to mimic a device profile operation). [example.log](https://github.com/thin-edge/thin-edge.io/files/15222902/example.log) is the full log, but below is just a snippet.

```text
==> /var/log/tedge/agent/workflow-device_profile-robot-123.log <==

==================================================================
Triggered device_profile workflow
==================================================================

topic:     te/device/main///cmd/device_profile/robot-123
operation: device_profile
cmd_id:    robot-123
time:      2024-05-06T12:19:40.883043984Z

==================================================================

----------------------[ device_profile @ init | time=2024-05-06T12:19:40.884932495Z ]----------------------

State:    {"profile":[{"operation":"firmware_update","payload":{"name":"core-image-tedge-rauc","remoteUrl":"https://t9679.latest.stage.c8y.io/inventory/binaries/82029","version":"20240430.1139"},"skip":false},{"operation":"software_update","payload":{"updateList":[{"modules":[{"name":"c8y-command-plugin","url":"","version":"latest"},{"name":"jq","url":"","version":"latest"}],"type":"apt"}]},"skip":false},{"operation":"config_update","payload":{"remoteUrl":"http://127.0.0.1:8001/c8y/inventory/binaries/35861751","type":"tedge-configuration-plugin"},"skip":true}],"status":"init"}

Action:   move to scheduled state


----------------------[ device_profile @ scheduled | time=2024-05-06T12:19:40.887326091Z ]----------------------

State:    {"profile":[{"operation":"firmware_update","payload":{"name":"core-image-tedge-rauc","remoteUrl":"https://t9679.latest.stage.c8y.io/inventory/binaries/82029","version":"20240430.1139"},"skip":false},{"operation":"software_update","payload":{"updateList":[{"modules":[{"name":"c8y-command-plugin","url":"","version":"latest"},{"name":"jq","url":"","version":"latest"}],"type":"apt"}]},"skip":false},{"operation":"config_update","payload":{"remoteUrl":"http://127.0.0.1:8001/c8y/inventory/binaries/35861751","type":"tedge-configuration-plugin"},"skip":true}],"status":"scheduled"}

Action:   /etc/tedge/operations/device_profile.sh scheduled '{"profile":[{"operation":"firmware_update","payload":{"name":"core-image-tedge-rauc","remoteUrl":"https://t9679.latest.stage.c8y.io/inventory/binaries/82029","version":"20240430.1139"},"skip":false},{"operation":"software_update","payload":{"updateList":[{"modules":[{"name":"c8y-command-plugin","url":"","version":"latest"},{"name":"jq","url":"","version":"latest"}],"type":"apt"}]},"skip":false},{"operation":"config_update","payload":{"remoteUrl":"http://127.0.0.1:8001/c8y/inventory/binaries/35861751","type":"tedge-configuration-plugin"},"skip":true}],"status":"scheduled"}'

Exit status: 0 (OK)

stderr <<EOF
Filtering/sorting device profile artifacts
EOF

stdout <<EOF
:::begin-tedge:::
{"currentIndex":-1,"profile":[{"operation":"firmware_update","payload":{"name":"core-image-tedge-rauc","remoteUrl":"https://t9679.latest.stage.c8y.io/inventory/binaries/82029","version":"20240430.1139"},"skip":false},{"operation":"software_update","payload":{"updateList":[{"modules":[{"name":"c8y-command-plugin","url":"","version":"latest"},{"name":"jq","url":"","version":"latest"}],"type":"apt"}]},"skip":false}]}
:::end-tedge:::
EOF


----------------------[ device_profile @ next_artifact | time=2024-05-06T12:19:40.913705569Z ]----------------------

State:    {"currentIndex":-1,"profile":[{"operation":"firmware_update","payload":{"name":"core-image-tedge-rauc","remoteUrl":"https://t9679.latest.stage.c8y.io/inventory/binaries/82029","version":"20240430.1139"},"skip":false},{"operation":"software_update","payload":{"updateList":[{"modules":[{"name":"c8y-command-plugin","url":"","version":"latest"},{"name":"jq","url":"","version":"latest"}],"type":"apt"}]},"skip":false}],"status":"next_artifact"}

Action:   /etc/tedge/operations/device_profile.sh next_artifact '{"currentIndex":-1,"profile":[{"operation":"firmware_update","payload":{"name":"core-image-tedge-rauc","remoteUrl":"https://t9679.latest.stage.c8y.io/inventory/binaries/82029","version":"20240430.1139"},"skip":false},{"operation":"software_update","payload":{"updateList":[{"modules":[{"name":"c8y-command-plugin","url":"","version":"latest"},{"name":"jq","url":"","version":"latest"}],"type":"apt"}]},"skip":false}],"status":"next_artifact"}'

Exit status: 0 (OK)

stderr <<EOF
Checking next artifact
Found next artifiact. status=firmware_update
EOF

stdout <<EOF
:::begin-tedge:::
{"status":"firmware_update","currentIndex":0,"current":{"operation":"firmware_update","payload":{"name":"core-image-tedge-rauc","remoteUrl":"https://t9679.latest.stage.c8y.io/inventory/binaries/82029","version":"20240430.1139"},"skip":false}}
:::end-tedge:::
EOF


----------------------[ device_profile @ firmware_update | time=2024-05-06T12:19:40.985850631Z ]----------------------

State:    {"current":{"operation":"firmware_update","payload":{"name":"core-image-tedge-rauc","remoteUrl":"https://t9679.latest.stage.c8y.io/inventory/binaries/82029","version":"20240430.1139"},"skip":false},"currentIndex":0,"profile":[{"operation":"firmware_update","payload":{"name":"core-image-tedge-rauc","remoteUrl":"https://t9679.latest.stage.c8y.io/inventory/binaries/82029","version":"20240430.1139"},"skip":false},{"operation":"software_update","payload":{"updateList":[{"modules":[{"name":"c8y-command-plugin","url":"","version":"latest"},{"name":"jq","url":"","version":"latest"}],"type":"apt"}]},"skip":false}],"status":"firmware_update"}

Action:   execute firmware_update as sub-operation


----------------------[ device_profile @ waiting_for_firmware_update | time=2024-05-06T12:19:40.98839402Z ]----------------------

State:    {"current":{"operation":"firmware_update","payload":{"name":"core-image-tedge-rauc","remoteUrl":"https://t9679.latest.stage.c8y.io/inventory/binaries/82029","version":"20240430.1139"},"skip":false},"currentIndex":0,"profile":[{"operation":"firmware_update","payload":{"name":"core-image-tedge-rauc","remoteUrl":"https://t9679.latest.stage.c8y.io/inventory/binaries/82029","version":"20240430.1139"},"skip":false},{"operation":"software_update","payload":{"updateList":[{"modules":[{"name":"c8y-command-plugin","url":"","version":"latest"},{"name":"jq","url":"","version":"latest"}],"type":"apt"}]},"skip":false}],"status":"waiting_for_firmware_update"}

Action:   await sub-operation completion


----------------------[ device_profile > firmware_update @ init | time=2024-05-06T12:19:41.002872349Z ]----------------------

State:    {"name":"core-image-tedge-rauc","remoteUrl":"https://t9679.latest.stage.c8y.io/inventory/binaries/82029","status":"init","version":"20240430.1139"}

Action:   move to scheduled state


----------------------[ device_profile > firmware_update @ scheduled | time=2024-05-06T12:19:41.004943235Z ]----------------------

State:    {"name":"core-image-tedge-rauc","remoteUrl":"https://t9679.latest.stage.c8y.io/inventory/binaries/82029","status":"scheduled","version":"20240430.1139"}

Action:   /usr/bin/sleep 1

Exit status: 0 (OK)

stderr (EMPTY)
stdout (EMPTY)
```

File: crates/common/logged_command/src/logged_command.rs [hunk 1]
```diff
@@ -227,21 +227,35 @@ impl LoggedCommand {
...
                         .await?
                 }
-                logger.write_all(b"stdout <<EOF\n").await?;
-                logger.write_all(&output.stdout).await?;
-                logger.write_all(b"EOF\n\n").await?;
-                logger.write_all(b"stderr <<EOF\n").await?;
-                logger.write_all(&output.stderr).await?;
-                logger.write_all(b"EOF\n").await?;
+                // Log stderr then stdout, so the flow reads chronologically
+                // as the stderr is used for log messages and the stdout is used for results
```
Comment (didier-wenzek) [inline crates/common/logged_command/src/logged_command.rs:241]: I don't really see why having `stderr` before `stdout` is better. But I have no strong argument for the reverse.

File: crates/core/tedge_agent/src/tedge_operation_converter/actor.rs [hunk 2]
```diff
@@ -560,13 +560,42 @@ impl CommandLog {
...
+==================================================================
+"#
+        );
+        if let Err(err) = self.write(&header_message).await {
+            error!("Fail to log to {}: {err}", self.path)
+        }
+    }
+
     async fn log_state_action(&mut self, state: &GenericCommandState, action: &OperationAction) {
+        if state.is_init() && !self.cmd_id.contains("sub:") {
```
Comment (didier-wenzek) [inline crates/core/tedge_agent/src/tedge_operation_converter/actor.rs]: I pushed a *fixup* so only `GenericCommandState` is aware of the structure of a command id.

File: crates/core/tedge_agent/src/tedge_operation_converter/actor.rs [hunk 3]
```diff
@@ -578,11 +607,26 @@ Action: {action}
...
+
+        let parent_operation = if cmd_id.starts_with("sub:") {
+            // Convert topic/payload: te/device/main///cmd/restart/sub:firmware_update:sub:device_profile:robot-123    {"status":"init"}
+            // into "device_profile > firmware_update > restart @ init"
+            let parents: Vec<&str> = cmd_id
+                .split("sub:")
+                .filter(|x| !x.is_empty())
+                .map(|x| x.split(':').next().unwrap())
+                .collect();
+            let parents: Vec<&str> = parents.into_iter().rev().collect();
```
Comment (didier-wenzek) [inline crates/core/tedge_agent/src/tedge_operation_converter/actor.rs]: I pushed a *fixup* so only `GenericCommandState` is aware of the structure of a command id.

**Review by didier-wenzek — APPROVED** [review]
Approved

**Pull request merged**